File: Source/Remote Control Server/ApiV3.vb
Imports System.Drawing
Imports System.Threading

Module ApiV3

    'API v3 uses commands similar to API v2 but supports communication from the server to an app

    Public Const COMMAND_IDENTIFIER As Byte = 127

    'first index
    Public Const cmd_connection As Byte = 10
    Public Const cmd_disconnect As Byte = 11
    Public Const cmd_pause As Byte = 12
    Public Const cmd_resume As Byte = 13
    Public Const cmd_control As Byte = 14
    Public Const cmd_pin As Byte = 15
    Public Const cmd_broadcast As Byte = 16
    Public Const cmd_set As Byte = 17
    Public Const cmd_get As Byte = 19
    Public Const cmd_open As Byte = 18


    'second index
    'connect
    Public Const cmd_connection_reachable As Byte = 0
    Public Const cmd_connection_protected As Byte = 1
    Public Const cmd_connection_connect As Byte = 2
    Public Const cmd_connection_disconnect As Byte = 3

    'action
    Public Const cmd_action_up As Byte = 0
    Public Const cmd_action_down As Byte = 1
    Public Const cmd_action_click As Byte = 2
    Public Const cmd_action_process As Byte = 3

    'setter
    Public Const cmd_set_pin As Byte = 0
    Public Const cmd_set_app_version As Byte = 1
    Public Const cmd_set_app_name As Byte = 2
    Public Const cmd_set_os_version As Byte = 3
    Public Const cmd_set_device_name As Byte = 5

    'getter
    Public Const cmd_get_server_version As Byte = 1
    Public Const cmd_get_server_name As Byte = 2
    Public Const cmd_get_os_name As Byte = 4
    Public Const cmd_get_screenshot As Byte = 5
    Public Const cmd_get_screenshots As Byte = 7
    Public Const cmd_get_api_version As Byte = 6

    'mouse
    Public Const cmd_mouse As Byte = 20
    Public Const cmd_mouse_pointers As Byte = 1
    Public Const cmd_mouse_pointers_absolute As Byte = 0
    Public Const cmd_mouse_pointers_absolute_presenter As Byte = 5
    Public Const cmd_mouse_pad_action As Byte = 2
    Public Const cmd_mouse_left_action As Byte = 3
    Public Const cmd_mouse_right_action As Byte = 4

    'keyboard
    Public Const cmd_keyboard As Byte = 21
    Public Const cmd_keyboard_unicode As Byte = 0
    Public Const cmd_keyboard_string As Byte = 1
    Public Const cmd_keyboard_keycode As Byte = 2


    Public Function isBroadcast(ByVal command As Command) As Boolean
        If command.data.Equals(New Byte() {COMMAND_IDENTIFIER, cmd_broadcast}) Then
            Return True
        ElseIf command.data(2) = ApiV3.cmd_connection_reachable Then
            Return True
        Else
            Return False
        End If
    End Function

    Public Function isConnectionCommand(ByVal command As Command) As Boolean
        Select Case command.data(1)
            Case cmd_connection
                Return True
            Case cmd_set
                Return True
        End Select
        Return False
    End Function

    Public Sub requestPin(ByRef app As App)
        'Send pin request
        Dim command As New Command
        command.source = Network.getServerIp()
        command.destination = app.ip
        command.priority = RemoteControlServer.Command.PRIORITY_HIGH

        Dim serverName As Byte() = New Byte() {COMMAND_IDENTIFIER, cmd_connection, cmd_connection_protected}
        command.data = buildCommandData(serverName, Converter.stringToByte(Server.gui.userName))

        command.send()
    End Sub

    Public Sub validatePin(ByRef app As App)
        'Send if pin is valid or not
        Dim command As New Command
        command.source = Network.getServerIp()
        command.destination = app.ip
        command.priority = RemoteControlServer.Command.PRIORITY_HIGH

        If (isAuthenticated(app.ip, app.pin)) Then
            command.data = New Byte() {COMMAND_IDENTIFIER, cmd_connection, cmd_connection_connect}
        Else
            command.data = New Byte() {COMMAND_IDENTIFIER, cmd_connection, cmd_connection_disconnect}
        End If

        command.send()
    End Sub

    Public Sub answerBroadCast(ByRef app As App)
        'Send server name back
        Dim command As New Command
        command.source = Network.getServerIp()
        command.destination = app.ip
        command.priority = RemoteControlServer.Command.PRIORITY_HIGH

        Dim serverName As Byte() = New Byte() {COMMAND_IDENTIFIER, cmd_get, cmd_get_server_name}
        command.data = buildCommandData(serverName, Converter.stringToByte(Server.gui.userName))

        command.send()
    End Sub

    Public Sub refuseBroadCast(ByRef app As App)

    End Sub

    Public Sub parseCommand(ByRef command As Command)
        Dim localCommand As Command = command
        Dim parseThread As Thread = New Thread(Sub() parseCommandThread(localCommand))
        parseThread.Start()
    End Sub

    Public Sub parseCommandThread(ByRef command As Command)
        Select Case command.data(1)
            Case cmd_connection
                parseConnectCommand(command)
            Case cmd_get
                answerGetRequest(command)
            Case cmd_set
                setValue(command)
            Case cmd_mouse
                parseMouseCommand(command)
            Case cmd_keyboard
                parseKeyboardCommand(command)
            Case cmd_open
                parseOpenCommand(command)
            Case Else
                Logger.add("Unknown command: " & Converter.commandToString(command))
        End Select
    End Sub

#Region "Connecting"

    Public Sub parseConnectCommand(ByRef command As Command)
        Dim app As App = Server.getApp(command.source)

        Select Case command.data(2)
            Case cmd_connection_connect
                'App wants to connect with the server       
                app.onConnect()
            Case cmd_connection_disconnect
                'App disconnected from the server       
                app.onDisconnect()
            Case cmd_connection_reachable
                'App checks if server is reachable
                'Reply with the server name

                If command.data.Length > 2 Then
                    'Command data contains source IP
                    app.ip = Converter.byteToString(command.data, 3)
                End If

                'answerBroadCast(app)
                app.onBroadCast(command)
                Logger.add(app.ip & " checked reachability")
            Case Else
                Logger.add("Unknown connection command: " & Converter.commandToString(command))
        End Select
    End Sub

#End Region

#Region "Open"

    Public Sub parseOpenCommand(ByRef command As Command)
        Dim app As App = Server.getApp(command.source)

        Select Case command.data(2)
            Case cmd_action_process
                Dim processString As String = Converter.byteToString(command.data, 3)
                Logger.add("Starting process: " & processString)
                Try
                    Process.Start(processString)
                Catch ex As Exception
                    Logger.add(ex.Message)
                End Try
            Case Else
                Logger.add("Unknown open command: " & Converter.commandToString(command))
        End Select
    End Sub

#End Region

#Region "Get and Set requests"

    Public Sub answerGetRequest(ByRef requestCommand As Command)
        Dim app As App = Server.getApp(requestCommand.source)

        Dim responseCommand As New Command
        responseCommand.source = Network.getServerIp()
        responseCommand.destination = app.ip
        responseCommand.priority = RemoteControlServer.Command.PRIORITY_HIGH

        Dim commandIdentifier As Byte() = New Byte() {COMMAND_IDENTIFIER, cmd_get, requestCommand.data(2)}

        Select Case requestCommand.data(2)
            Case cmd_get_server_version
                'responseCommand.data = buildCommandData(commandIdentifier, Converter.stringToByte(Server.getServerVersionName))
                responseCommand.data = buildCommandData(commandIdentifier, New Byte() {Updater.currentVersionCode})
            Case cmd_get_server_name
                responseCommand.data = buildCommandData(commandIdentifier, Converter.stringToByte(Server.gui.userName))
            Case cmd_get_os_name
                responseCommand.data = buildCommandData(commandIdentifier, Converter.stringToByte(Server.getServerOs))
            Case cmd_get_api_version
                responseCommand.data = buildCommandData(commandIdentifier, New Byte() {3})
            Case cmd_get_screenshot
                parseScreenshotProperties(requestCommand)
                answerScreenGetRequest(requestCommand, responseCommand)
            Case cmd_get_screenshots
                parseScreenshotProperties(requestCommand)
                Screenshot.continueSendingScreenshots = True
                If Not Screenshot.isSendingScreenshot Then
                    Screenshot.keepSendingScreenshots(requestCommand, responseCommand)
                End If
            Case Else
                Logger.add("Unknown get command: " & Converter.commandToString(requestCommand))
        End Select

        If Not responseCommand.data Is Nothing Then
            responseCommand.send()
        End If
    End Sub

    Public Sub parseScreenshotProperties(ByVal requestCommand As Command)
        If requestCommand.data.Length > 3 Then
            Screenshot.lastRequestedWidth = requestCommand.data(3) * 100

            If requestCommand.data.Length > 4 Then
                Screenshot.lastRequestedQuality = requestCommand.data(4)
            Else
                Screenshot.lastRequestedQuality = Settings.screenQualityFull
            End If
        Else
            Screenshot.lastRequestedWidth = 9999
        End If

        'Logger.add("Screenshot requested. Width: " & Screenshot.lastRequestedWidth & " Quality: " & Screenshot.lastRequestedQuality)
    End Sub

    Public Sub answerScreenGetRequest(ByVal requestCommand As Command, ByVal responseCommand As Command)
        'Logger.add("Sending screenshot. Width: " & Screenshot.lastRequestedWidth & " Quality: " & Screenshot.lastRequestedQuality)

        Dim screenshotBitmap As Bitmap = Screenshot.getResizedScreenshot(Screenshot.lastRequestedWidth)
        Dim screenshotData As Byte() = Converter.bitmapToByte(screenshotBitmap, Screenshot.lastRequestedQuality)
        Dim commandIdentifier = New Byte() {COMMAND_IDENTIFIER, cmd_get, cmd_get_screenshot, Screenshot.lastRequestedWidth / 100}
        responseCommand.data = buildCommandData(commandIdentifier, screenshotData)
        responseCommand.priority = Command.PRIORITY_MEDIUM
        responseCommand.send()
    End Sub

    Public Sub setValue(ByRef setCommand As Command)
        Dim app As App = Server.getApp(setCommand.source)

        Select Case setCommand.data(2)
            Case cmd_set_pin
                app.pin = Converter.byteToString(setCommand.data, 3)
                Logger.add("Pin from " & app.ip & " set to " & app.pin)
                validatePin(app)
            Case cmd_set_app_version
                app.appVersion = Converter.byteToString(setCommand.data, 3)
                Logger.add("App version from " & app.ip & " set to " & app.appVersion)
            Case cmd_set_app_name
                app.appName = Converter.byteToString(setCommand.data, 3)
                Logger.add("App name from " & app.ip & " set to " & app.appName)
            Case cmd_set_os_version
                app.osVersion = Converter.byteToString(setCommand.data, 3)
                app.detectOs()
                Logger.add("OS version from " & app.ip & " set to " & app.osVersion)
            Case cmd_set_device_name
                app.deviceName = Converter.byteToString(setCommand.data, 3)
                Logger.add("Device name from " & app.ip & " set to " & app.deviceName)
            Case Else
                Logger.add("Unknown set command: " & Converter.commandToString(setCommand))
        End Select

    End Sub

#End Region

#Region "Mouse & Keyboard"

    Public Sub parseMouseCommand(ByRef command As Command)
        Dim app As App = Server.getApp(command.source)

        Select Case command.data(2)
            Case cmd_mouse_pointers
                MouseV3.parsePointerData(command.data)
            Case cmd_mouse_pointers_absolute
                MouseV3.parseAbsolutePointerData(command.data, False)
            Case cmd_mouse_pointers_absolute_presenter
                MouseV3.parseAbsolutePointerData(command.data, True)
            Case cmd_mouse_pad_action
                Select Case command.data(3)
                    Case cmd_action_down
                        MouseV3.pointersDown()
                    Case cmd_action_up
                        MouseV3.pointersUp()
                    Case Else
                        Logger.add("Unknown mouse pad command")
                End Select
            Case cmd_mouse_left_action
                Select Case command.data(3)
                    Case cmd_action_down
                        Logger.add("Mouse left down")
                        MouseV3.leftMouseDown()
                    Case cmd_action_up
                        Logger.add("Mouse left up")
                        MouseV3.leftMouseUp()
                    Case Else
                        Logger.add("Unknown mouse left command")
                End Select
            Case cmd_mouse_right_action
                Select Case command.data(3)
                    Case cmd_action_down
                        Logger.add("Mouse right down")
                        MouseV3.rightMouseDown()
                    Case cmd_action_up
                        Logger.add("Mouse right up")
                        MouseV3.rightMouseUp()
                    Case Else
                        Logger.add("Unknown mouse right command")
                End Select
            Case Else
                Logger.add("Unknown mouse command: " & Converter.commandToString(command))
        End Select
    End Sub

    Public Sub parseKeyboardCommand(ByRef command As Command)
        Dim app As App = Server.getApp(command.source)

        Select Case command.data(2)
            Case cmd_keyboard_keycode
                Dim action As Byte = command.data(3)
                Dim keyCode As Integer = command.data(6)
                keyCode = (keyCode << 8) + command.data(7)
                'Logger.add("Received key code: " & keyCode)

                If keyCode < Keyboard.KEYCODE_C1 Or keyCode > Keyboard.KEYCODE_C12 Then
                    Select Case action
                        Case cmd_action_down
                            Keyboard.sendKeyDown(Keyboard.keycodeToKey(keyCode))
                        Case cmd_action_up
                            Keyboard.sendKeyUp(Keyboard.keycodeToKey(keyCode))
                        Case cmd_action_click
                            Dim keyboardKey As Key = Keyboard.keycodeToKey(keyCode)
                            If keyboardKey = Keyboard.KEYCODE_UNKOWN Then
                                Keyboard.keycodeToShortcut(keyCode)
                            Else
                                Keyboard.sendKeyPress(keyboardKey)
                            End If
                    End Select
                Else
                    'It's a custom key, get action from config
                    Dim actionIndex = keyCode - Keyboard.KEYCODE_C1
                    Logger.add("Received custom key code: " & (actionIndex + 1))

                    If actionIndex > Settings.customActions.Count - 1 Then
                        Logger.add("No custom action set")
                        Process.Start("http://remote-control-collection.com/help/custom/")
                        Logger.trackEevent("Server", "Custom", "Not set")
                    Else
                        Try
                            Dim path As String = Settings.customActions(actionIndex)
                            Process.Start(path)
                            Logger.trackEevent("Server", "Custom", path)
                        Catch ex As Exception
                            Logger.add("Unable to invoke custom action:")
                            Logger.add(ex.ToString)
                        End Try
                    End If


                End If
            Case cmd_keyboard_string
                Dim keyString As String = Converter.byteToString(command.data, 3)
                'Logger.add("Received key string: " & keyString)
                Keyboard.sendEachKey(keyString)
            Case Else
                Logger.add("Unknown keyboard command: " & Converter.commandToString(command))
        End Select

    End Sub

#End Region

End Module

File: Source/MacOS/RemoteControlServer/apiv3.cpp
#include "apiv3.h"
#include "app.h"
#include "authentication.h"
#include "converter.h"
#include "logger.h"
#include "mousev3.h"
#include "network.h"
#include "screenshot.h"
#include "serial.h"
#include "server.h"
#include "settings.h"

#ifdef Q_OS_MAC
    #include "keyboardmac.h"
    #include "mousemac.h"
#endif

#include <QByteArray>
#include <QProcess>
#include <QtConcurrent>

#include <QDebug>

ApiV3* ApiV3::instance = NULL;

ApiV3* ApiV3::Instance()
{
    if (!instance)
    {
        instance = new ApiV3();
    }
    return instance;
}

ApiV3::ApiV3() :
    COMMAND_IDENTIFIER(127)
{
}

bool ApiV3::isBroadcast(Command &command)
{
    return ((command.data == QByteArray({COMMAND_IDENTIFIER, cmd_broadcast}))
             || (command.data->length() >= 3 && command.data->at(2) == cmd_connection_reachable))
            ? true : false;
}

bool ApiV3::isConnectionCommand(Command &command)
{
    return (command.data->length() >= 2 && (command.data->at(1) == cmd_connection || command.data->at(1) == cmd_set))
            ? true : false;
}

void ApiV3::requestPin(App &app)
{
    Command *command = new Command();
    command->source = Network::Instance()->getServerIp();
    command->destination = app.ip;
    command->priority = Command::PRIORITY_HIGH;
    QByteArray *serverName = new QByteArray();
    serverName->append(COMMAND_IDENTIFIER).append(cmd_connection).append(cmd_connection_protected);
    serverName->append(Server::Instance()->userName.toUtf8());
    command->data = serverName;
    command->send();
}

void ApiV3::validatePin(App &app)
{
    Command *command = new Command();
    command->source = Network::Instance()->getServerIp();
    command->destination = app.ip;
    command->priority = Command::PRIORITY_HIGH;
    QByteArray *commandData = new QByteArray();
    commandData->append(COMMAND_IDENTIFIER).append(cmd_connection);
    if(Authentication::Instance()->isAuthenticated(app.ip, app.pin))
    {
        commandData->append(cmd_connection_connect);
    }
    else
    {
        commandData->append(cmd_connection_disconnect);
    }
    command->data = commandData;
    command->send();
}

void ApiV3::answerBroadCast(App &app)
{
    Command *command = new Command();
    command->source = Network::Instance()->getServerIp();
    command->destination = app.ip;
    command->priority = Command::PRIORITY_HIGH;
    QByteArray *serverName = new QByteArray();
    serverName->append(COMMAND_IDENTIFIER).append(cmd_get).append(cmd_get_server_name);
    serverName->append(Server::Instance()->userName.toUtf8());
    command->data = serverName;
    command->send();
}

void ApiV3::parseCommand(Command &command)
{
    if (command.data->length() >= 2)
    {
        switch (command.data->at(1))
        {
        case cmd_connection:
            parseConnectCommand(command);
            break;
        case cmd_get:
            answerGetRequest(command);
            break;
        case cmd_set:
            setValue(command);
            break;
        case cmd_mouse:
            parseMouseCommand(command);
            break;
        case cmd_keyboard:
            parseKeyboardCommand(command);
            break;
        case cmd_open:
            parseOpenCommand(command);
            break;
        default:
            Logger::Instance()->add("Unknown command " + Converter::Instance()->commandToString(command));
            break;
        }
    }
}

void ApiV3::parseConnectCommand(Command &command)
{
    App *app = Server::Instance()->getApp(command.source);
    if (command.data->length() >= 3)
    {
        switch (command.data->at(2))
        {
        case cmd_connection_connect:
            app->onConnect();
            break;
        case cmd_connection_disconnect:
            app->onDisconnect();
            break;
        case cmd_connection_reachable:
            app->ip = Converter::Instance()->byteToString(*command.data, 3);
            app->onBroadCast(command);
            Logger::Instance()->add(app->ip + " checked reachability");
            break;
        default:
            Logger::Instance()->add("Unknown connection command: " + Converter::Instance()->commandToString(command));
            break;
        }
    }
}

void ApiV3::parseOpenCommand(Command &command)
{
    App *app = Server::Instance()->getApp(command.source);
    if (command.data->length() >= 3 && command.data->at(2) == cmd_action_process)
    {
        QString processString = Converter::Instance()->byteToString(*command.data, 3);
        Logger::Instance()->add("Starting process: " + processString);
        Server::Instance()->startProcess(processString);
    }
    else
    {
        Logger::Instance()->add("Unknown open command: " + Converter::Instance()->commandToString(command));
    }
}

void ApiV3::answerGetRequest(Command &requestCommand)
{
    App *app = Server::Instance()->getApp(requestCommand.source);
    Command *responseCommand = new Command();
    responseCommand->source = Network::Instance()->getServerIp();
    responseCommand->destination = app->ip;
    responseCommand->priority = Command::PRIORITY_HIGH;

    if (requestCommand.data->length() >= 3)
    {
        QByteArray *commandIdentifier = new QByteArray();
        commandIdentifier->append(COMMAND_IDENTIFIER).append(cmd_get).append(requestCommand.data->at(2));

        switch (requestCommand.data->at(2))
        {
        case cmd_get_server_version:
            Logger::Instance()->add("Server version requested");
            responseCommand->data = &commandIdentifier->append(APP_VERSION); // currentVersionCode from update requested
            break;
        case cmd_get_server_name:
            responseCommand->data = &commandIdentifier->append(Server::Instance()->userName.toUtf8());
            break;
        case cmd_get_os_name:
            responseCommand->data = &commandIdentifier->append(Server::Instance()->getServerOs().toUtf8());
            break;
        case cmd_get_api_version:
            responseCommand->data = &commandIdentifier->append(3);
            break;
        case cmd_get_screenshot:
            parseScreenshotProperties(requestCommand);
            answerScreenGetRequest(*responseCommand);
            break;
        case cmd_get_screenshots:
            parseScreenshotProperties(requestCommand);
            Screenshot::Instance()->continueSendingScreenshots = true;
            if (!Screenshot::Instance()->isSendingScreenshot)
            {
                Screenshot::Instance()->keepSendingScreenshots(requestCommand, *responseCommand);
            }
            break;
        default:
            Logger::Instance()->add("Unknown get command: " + Converter::Instance()->commandToString(requestCommand));
            break;
        }
    }

    if (responseCommand->data->length() > 0)
    {
        responseCommand->send();
    }
}

void ApiV3::parseScreenshotProperties(Command &command)
{
    if (command.data->length() >= 4)
    {
        Screenshot::Instance()->lastRequestedWidth = (unsigned char)command.data->at(3) * 100;

        if (command.data->length() >= 5)
        {
            Screenshot::Instance()->lastRequestedQuality = (unsigned char)command.data->at(4);
        }
        else
        {
            Screenshot::Instance()->lastRequestedQuality = Settings::Instance()->screenQualityFull;
        }
    }
    else
    {
        Screenshot::Instance()->lastRequestedWidth = 9999;
    }
}

void ApiV3::answerScreenGetRequest(Command &responseCommand)
{
    QPixmap* screenshotBitmap = Screenshot::Instance()->getResizedScreenshot(Screenshot::Instance()->lastRequestedWidth);
    QByteArray *screenshotData = Converter::Instance()->bitmapToByte(*screenshotBitmap, Screenshot::Instance()->lastRequestedQuality);
    QByteArray *commandIdentifier = new QByteArray();
    commandIdentifier->append(COMMAND_IDENTIFIER).append(cmd_get).append(cmd_get_screenshot);
    commandIdentifier->append(Screenshot::Instance()->lastRequestedWidth / 100);
    responseCommand.data = commandIdentifier;
    responseCommand.data->append(*screenshotData);
    responseCommand.send();
}

void ApiV3::setValue(Command &setCommand)
{
    App *app = Server::Instance()->getApp(setCommand.source);
    if (setCommand.data->length() >= 3)
    {
        switch (setCommand.data->at(2))
        {
        case cmd_set_pin:
            app->pin = Converter::Instance()->byteToString(*setCommand.data, 3);
            Logger::Instance()->add("Pin from " + app->ip + " set to " + app->pin);
            validatePin(*app);
            break;
        case cmd_set_app_version:
            app->appVersion = Converter::Instance()->byteToString(*setCommand.data, 3);
            Logger::Instance()->add("App version from " + app->ip + " set to " + app->appVersion);
            break;
        case cmd_set_app_name:
            app->appName = Converter::Instance()->byteToString(*setCommand.data, 3);
            Logger::Instance()->add("App name from " + app->ip + " set to " + app->appName);
            break;
        case cmd_set_os_version:
            app->osVersion = Converter::Instance()->byteToString(*setCommand.data, 3);
            Logger::Instance()->add("Os version from " + app->ip + " set to " + app->osVersion);
            break;
        case cmd_set_device_name:
            app->deviceName = Converter::Instance()->byteToString(*setCommand.data, 3);
            Logger::Instance()->add("Device name from " + app->ip + " set to " + app->deviceName);
            break;
        default:
            Logger::Instance()->add("Unknown set command: " + Converter::Instance()->commandToString(setCommand));
            break;
        }
    }
}

void ApiV3::parseMouseCommand(Command &command)
{
    if (command.data->length() >= 3)
    {
        switch (command.data->at(2))
        {
        case cmd_mouse_pointers:
            MouseV3::Instance()->parsePointerData(*command.data);
            break;
        case cmd_mouse_pointers_absolute:
            MouseV3::Instance()->parseAbsolutePointerData(*command.data, false);
            break;
        case cmd_mouse_pointers_absolute_presenter:
            MouseV3::Instance()->parseAbsolutePointerData(*command.data, true);
            break;
        case cmd_mouse_pad_action:
            if (command.data->length() >= 4)
            {
                switch (command.data->at(3))
                {
                case cmd_action_down:
                    MouseV3::Instance()->pointersDown();
                    break;
                case cmd_action_up:
                    MouseV3::Instance()->pointersUp();
                    break;
                default:
                    Logger::Instance()->add("Unknown mouse pad command");
                    break;
                }
            }
            break;
        case cmd_mouse_left_action:
            if (command.data->length() >= 4)
            {
                switch (command.data->at(3))
                {
                case cmd_action_down:
                    Logger::Instance()->add("Mouse left down");
                    MouseMac::Instance()->leftMouseDown(false);
                    break;
                case cmd_action_up:
                    Logger::Instance()->add("Mouse left up");
                    MouseMac::Instance()->leftMouseUp(false);
                    break;
                default:
                    Logger::Instance()->add("Unknown mouse left command");
                    break;
                }
            }
            break;
        case cmd_mouse_right_action:
            if (command.data->length() >= 4)
            {
                switch (command.data->at(3))
                {
                case cmd_action_down:
                    Logger::Instance()->add("Mouse right down");
                    MouseMac::Instance()->rightMouseDown();
                    break;
                case cmd_action_up:
                    Logger::Instance()->add("Mouse right up");
                    MouseMac::Instance()->rightMouseUp();
                    break;
                default:
                    Logger::Instance()->add("Unknown mouse right command");
                    break;
                }
            }
            break;
        default:
            Logger::Instance()->add("Unknown mouse command: " + Converter::Instance()->commandToString(command));
            break;
        }
    }
}

void ApiV3::parseKeyboardCommand(Command &command)
{
    if (command.data->length() >= 3)
    {
        switch (command.data->at(2))
        {
        case cmd_keybaord_keycode:
        {
            if (command.data->length() >= 8)
            {
                char action = command.data->at(3);
                unsigned char firstByte = command.data->at(6);
                unsigned char secondByte = command.data->at(7);
                int keyCode = firstByte << 8 | secondByte;

#ifdef Q_OS_MAC
                if (keyCode < KeyboardMac::KEYCODE_C1 || keyCode > KeyboardMac::KEYCODE_C12)
#endif
                {
                    switch (action)
                    {
                    case cmd_action_down:
#ifdef Q_OS_MAC
                        KeyboardMac::Instance()->sendKeyDown(KeyboardMac::Instance()->keycodeToKey(keyCode));
#endif
                        break;
                    case cmd_action_up:
#ifdef Q_OS_MAC
                        KeyboardMac::Instance()->sendKeyUp(KeyboardMac::Instance()->keycodeToKey(keyCode));
#endif
                        break;
                    case cmd_action_click:
#ifdef Q_OS_MAC
                        CGKeyCode keyboardKey = KeyboardMac::Instance()->keycodeToKey(keyCode);
                        if (keyboardKey == KeyboardMac::KEYCODE_UNKOWN)
                        {
                            KeyboardMac::Instance()->sendShortcut(keyboardKey);
                        }
                        else
                        {
                            KeyboardMac::Instance()->sendKeyPress(keyboardKey);
                        }
#endif
                        break;
                    }
                }
                else
                {
#ifdef Q_OS_MAC
                    int actionIndex = keyCode - KeyboardMac::KEYCODE_C1;
#endif
                    Logger::Instance()->add("Received custom key code: " + (actionIndex + 1));

                    if (Settings::Instance()->serialCommands)
                    {
                        Serial::Instance()->sendCommand(command);
                    }
                    else
                    {
                        if (actionIndex > Settings::Instance()->customActions->length() - 1)
                        {
                            Logger::Instance()->add("No custom action set");
                            Server::Instance()->startProcess("http://remote-control-collection.com/help/custom/");
                            Logger::Instance()->trackEvent("Server", "Custom", "Not set");
                        }
                        else
                        {
                            QString path = Settings::Instance()->customActions->at(actionIndex);
                            Server::Instance()->startProcess(path);
                            Logger::Instance()->trackEvent("Server", "Custom", path);
                        }
                    }
                }
            }
            break;
        }
        case cmd_keyboard_string:
        {
            QString keyString = Converter::Instance()->byteToString(*command.data, 3);
#ifdef Q_OS_MAC
            KeyboardMac::Instance()->sendEachKey(keyString);
#endif
            break;
        }
        default:
        {
            Logger::Instance()->add("Unknown keyboard command: " + Converter::Instance()->commandToString(command));
            break;
        }
        }
    }
}
